lib/msal-common/src/crypto/PopTokenGenerator.ts

```diff
@@ -37,31 +37,42 @@ export class PopTokenGenerator {
     }
 
     async generateCnf(resourceRequestMethod: string, resourceRequestUri: string): Promise<string> {
+        const reqCnf = await this.generateKid(resourceRequestMethod, resourceRequestUri);
+        return this.cryptoUtils.base64Encode(JSON.stringify(reqCnf));
+    }
+
+    async generateKid(resourceRequestMethod: string, resourceRequestUri: string): Promise<ReqCnf> {
         const kidThumbprint = await this.cryptoUtils.getPublicKeyThumbprint(resourceRequestMethod, resourceRequestUri);
-        const reqCnf: ReqCnf = {
+
+        return {
             kid: kidThumbprint,
             xms_ksl: KeyLocation.SW
         };
-        return this.cryptoUtils.base64Encode(JSON.stringify(reqCnf));
     }
 
     async signPopToken(accessToken: string, resourceRequestMethod: string, resourceRequestUri: string): Promise<string> {
         const tokenClaims: TokenClaims | null = AuthToken.extractTokenClaims(accessToken, this.cryptoUtils);
-        const resourceUrlString: UrlString = new UrlString(resourceRequestUri);
-        const resourceUrlComponents: IUri = resourceUrlString.getUrlComponents();
-
+        
         if (!tokenClaims?.cnf?.kid) {
             throw ClientAuthError.createTokenClaimsRequiredError();
         }
 
+        return this.signPayload(accessToken, tokenClaims.cnf.kid, resourceRequestMethod, resourceRequestUri);
+    }
+
+    async signPayload(payload: string, kid: string, resourceRequestMethod: string, resourceRequestUri: string, claims?: object): Promise<string> {
+        const resourceUrlString: UrlString = new UrlString(resourceRequestUri);
+        const resourceUrlComponents: IUri = resourceUrlString.getUrlComponents();
+
         return await this.cryptoUtils.signJwt({
-            at: accessToken,
+            at: payload,
```

Comment:
Maybe call it something other than "payload"

Reply:
We could call it "secret". For internal proof-of-possession its a good way to distinguish the access token secret from the "PoP AT". Would "secret" accurately describe any opaque string though?

Reply:
Yeah `secret` works.

Reply:
Going to keep as payload, secret doesn't seem appropriate semantically.